includes/class-amp-theme-support.php

```diff
@@ -303,13 +337,246 @@ public static function is_customize_preview_iframe() {
+				$callback = $id;
+			} else {
+				$callback = $supportable_template['callback'];
+			}
+
+			// @todo Consider using the $id as the conditional callback.
+			// If the available_callback is a method on the query, then call the method on the query itself.
+			if ( is_string( $callback ) && 'is_' === substr( $callback, 0, 3 ) && method_exists( $query, $callback ) ) {
+				$is_match = call_user_func( array( $query, $callback ) );
+			} else {
+				$is_match = call_user_func( $callback );
+			}
+
+			if ( $is_match ) {
+				$matching_templates[ $id ] = (
+					! empty( $supportable_template['supported'] )
+					||
+					( AMP_Options_Manager::get_option( 'all_templates_supported' ) && empty( $supportable_template['immutable'] ) )
+				);
+			}
+		}
+
+		// Make sure children override their parents.
+		$matching_template_ids = array_keys( $matching_templates );
+		foreach ( $matching_template_ids as $id ) {
+			$has_children = false;
+			foreach ( $supportable_templates as $other_id => $supportable_template ) {
+				if ( $other_id === $id ) {
+					continue;
+				}
+				if ( isset( $supportable_template['parent'] ) && $id === $supportable_template['parent'] ) {
+					$has_children = true;
+					break;
+				}
+			}
+
+			// Delete all matching parent templates since the child will override them.
+			if ( ! $has_children ) {
+				$supportable_template = $supportable_templates[ $id ];
+				while ( ! empty( $supportable_template['parent'] ) ) {
+					$parent               = $supportable_template['parent'];
+					$supportable_template = $supportable_templates[ $parent ];
+
+					// Let the child supported status override the parent's supported status.
+					unset( $matching_templates[ $parent ] );
+				}
+			}
+		}
+
+		// If there aren't any matching templates left that are supported, then we consider it to not be available.
+		if ( ! in_array( true, $matching_templates, true ) ) {
+			return new WP_Error( 'no_supported_template' );
+		}
+
+		// For singular queries, post_supports_amp() is given the final say.
+		if ( $query->is_singular() || $query->is_posts_page ) {
+			/**
+			 * Queried object.
+			 *
+			 * @var WP_Post $queried_object
+			 */
+			$queried_object = $query->get_queried_object();
+			$support_errors = AMP_Post_Type_Support::get_support_errors( $queried_object );
+			if ( ! empty( $support_errors ) ) {
+				$error = new WP_Error();
+				foreach ( $support_errors as $support_error ) {
+					$error->add( $support_error, '' );
+				}
+				return $error;
+			}
+		}
+
+		// If all checks have passed, then the template is available.
+		return true;
+	}
+
+	/**
+	 * Get the templates which can be supported.
+	 *
+	 * @return array Supportable templates.
+	 */
+	public static function get_supportable_templates() {
+		$templates = array(
+			'is_singular' => array(
+				'label'       => __( 'Singular', 'amp' ),
+				'description' => __( 'Required for the above content types.', 'amp' ),
```

Comment:
<img width="728" alt="clicking-content-type" src="https://user-images.githubusercontent.com/4063887/42111662-c7e23e16-7baa-11e8-9689-9d12c9c988e3.png">

Reply:
@kienstra I like the idea of making it simpler for people, so that if they select a content type the singular template would be auto selected.

Is there ever a situation where I would select a singular template and NOT select a content type?

Reply:
@jwold Yes, it would probably be common to have the Singular template checked but not have all Content Types checked. What would probably not be as common is to have Singular unchecked but the Content Types checked. There is a scenario where it could be used though: maybe the post archive page for a given CPT can be served as AMP but the singular CPT template cannot. That would surely be rare, but even so the “Required“ comment above isn't really true.